---
title: "10/28/2021 Generating a finalized lineage analysis script for Zhang Lab/CSHL poster"
output: html_notebook
---

# Initialize the data
```{R include=FALSE}
rm(list = ls())

library(dplyr)
library(Seurat)
library(patchwork)
library(RColorBrewer)
library(ggplot2)
library(Biostrings)
library(dplyr)
library(stringr)
library(sjmisc)
library(gdata)
library(venn)
library(kit)
```

# Load in the gDNA barcode data
```{r include=FALSE}
temp <- list.files(path='/Volumes/Super_cells/2021_09_29_United_gDNA_10xcDNA_BCs/starcode_outputs', pattern='*gDNA*', full.names = T)
gDNA_files <- lapply(temp, read.delim, header = F)
list_names <- list.files(path='/Volumes/Super_cells/2021_09_29_United_gDNA_10xcDNA_BCs/starcode_outputs', pattern='*gDNA*')
names(gDNA_files) <- list_names

reference <- read.delim('/Volumes/Super_cells/2021_09_29_United_gDNA_10xcDNA_BCs/CellRanger_inputs/FeatureReference.csv', sep = ',')

startseq <- "GCTGTACAAGTAGGAT"
```

# Remove files no longer needed
```{r include=FALSE}
rm(temp, list_names)
```

# Give the columns from each df descriptive names
```{r}
for (i in names(gDNA_files)){
  names(gDNA_files[[i]]) <- c('sample','reads')
}

```

# Get basic statistics of each condition
```{r}
# Establish how many total barcodes per condition (should be the same since they have the same reference), number of total reads per condition, number pf barcodes with more than 0 reads, number of barcodes with more than 2 reads, number of reads from barcodes with more than 2 reads, and how many total reads are lost by filtering on barcodes with at least 2 reads
gDNA_basic_stats <- data.frame()
for (i in names(gDNA_files)){
  temp_df <- data.frame(Cond = i, Num_BCs = nrow(gDNA_files[[i]]), Num_reads = sum(gDNA_files[[i]]$reads), Num_BCs_greater0 = nrow(gDNA_files[[i]][gDNA_files[[i]]$reads > 0,]), Num_BCs_greater2 = nrow(gDNA_files[[i]][gDNA_files[[i]]$reads > 2,]),Num_reads_greater2 = sum(gDNA_files[[i]][gDNA_files[[i]]$reads > 2,]$reads), reads_lost = sum(gDNA_files[[i]]$reads)-sum(gDNA_files[[i]][gDNA_files[[i]]$reads > 2,]$reads))
  gDNA_basic_stats <- rbind(gDNA_basic_stats,temp_df)
}

rm(temp_df)
# all samples have the same number of barcodes to begin with, thats good as they were all run together
```

# Filter the reference barcode list to remove "bad" barcodes based on sequence
```{r}
reference_original <- reference # Retain the old list for comparison 

# Remove barcodes with ambiguous reads and that stray from "WSN" repeat (do this by removing runs of >= 4 which should not be possible)
reference <- reference[(unlist(lapply(reference$sequence, function(x){ grepl('N',x, fixed = T)}))==0),] 
reference <- reference[(unlist(lapply(reference$sequence, function(x){ grepl('AAAA',x, fixed = T)}))==0),]
reference <- reference[(unlist(lapply(reference$sequence, function(x){ grepl('TTTT',x, fixed = T)}))==0),]
reference <- reference[(unlist(lapply(reference$sequence, function(x){ grepl('CCCC',x, fixed = T)}))==0),]
reference <- reference[(unlist(lapply(reference$sequence, function(x){ grepl('GGGG',x, fixed = T)}))==0),] # This cleans from 703725 sequences down to 307561 sequences
```

# Generate rank ordered barcode plots - pre-filtering by reference
```{r}
plot_list <- list()
for (i in names(gDNA_files)){
  plot.new()
  plot_list[[i]] <- ggplot(data = gDNA_files[[i]][gDNA_files[[i]]$reads > 2,], aes(reorder(sample,-reads), log2(reads)))+geom_bar(stat='identity') +  
    theme(axis.text.x=element_blank()) + labs(x = 'Rank ordered barcodes', y = 'Log(reads per barcode)', title = paste(i, 'n =',length(gDNA_files[[i]][gDNA_files[[i]]$reads > 2,]$reads)))

}

# Display the plots
plot_list[[1]]
plot_list[[2]]
plot_list[[3]]
plot_list[[4]]
plot_list[[5]]
```
# Filter by reference list
```{r}
# allow R to use more memory
options(future.globals.maxSize= 35 * 1024^2)

# Build a boolean of which barcodes to keep by checking that the sample has the startseq and that the barcode sequence is in the updated reference file
boolean_present <- logical()
for (i in 1:nrow(gDNA_files[[1]])){ 
  if (i%%10000 == 0){
    print(i)
  }
  boolean_present[[i]] <- substr(gDNA_files$sc_output_counts_DLS005_Acid_gDNA_S36.txt$sample[i],1,16) == startseq & str_contains(reference$sequence,substr(gDNA_files$sc_output_counts_DLS005_Acid_gDNA_S36.txt$sample[i],17,86))
}

save.image(file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/workspace_ref_filt.RData')
```

# Load checkpoint of post reference filtering
```{r}
load('/Volumes/Super_cells/2021_10_28_Cleaned_Data/workspace_ref_filt.RData')
```

# Use boolean present to filter for just barcodes with good sequences
```{r include=FALSE}
gDNA_filt <- list()
for (i in names(gDNA_files)){
  gDNA_filt[[i]] <- gDNA_files[[i]][boolean_present,]
}
```

# Generate rank ordered barcode plots after filtering
```{r include=FALSE}
plot_list <- list()
for (i in names(gDNA_filt)){
  plot.new()
  plot_list[[i]] <- ggplot(data = gDNA_filt[[i]][gDNA_filt[[i]]$reads > 2,], aes(reorder(sample,-reads), log2(reads)))+geom_bar(stat='identity') +
    theme(axis.text.x=element_blank()) + labs(x = 'Rank ordered barcodes', y = 'Log(reads per barcode)', title = paste0(i, "n = ", nrow(gDNA_filt[[i]][gDNA_filt[[i]]$reads > 2,])))
}

# Display the plots
plot_list[[1]]
plot_list[[2]]
plot_list[[3]]
plot_list[[4]]
plot_list[[5]]
```

# Assign the correct lineage name to each of the conditions
```{r}
gDNA_anno <- list()
for (j in names(gDNA_filt)){
  temp <- gDNA_filt[[j]]
  temp$sequence <- substr(temp$sample, 17, nchar(temp$sample))
  temp2 <- merge(temp, reference, by  = 'sequence')
  temp3 <- data.frame(Lineage = temp2$name, BC_seq = temp2$sequence, Full_seq = temp2$sample, Reads = temp2$reads)
  gDNA_anno[[j]] <- temp3[order(-temp3$Reads),]
}
```

# Looking at different thresholds for determining resistant barcodes and finding overlaps between different conditions
```{r}

# Establishing resistant barcodes as those that have >1000 reads in a gDNA condition
greater1000 <- list(Acid = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage[gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Reads > 1000],
                      Cis = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage[gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Reads > 1000],
                      CoCl2 = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage[gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Reads > 1000],
                      Dab = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage[gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Reads > 1000],
                      Tram = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage[gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Reads > 1000])
greater1000_venn <- venn(greater1000, intersections = T)
greater1000_venn_data <- attr(greater1000_venn,"intersections")

# Establish resistant lineages as the 50 barcodes with the highest expression
top50 <- list(Acid = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage[1:50],
                      Cis = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage[1:50],
                      CoCl2 = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage[1:50],
                      Dab = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage[1:50],
                      Tram = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage[1:50])
top50_venn <- venn(top50, intersections = T)
top50_venn_data <- attr(top50_venn,"intersections")

# Establish resistant lineages as the 100 barcodes with the highest expression
top100 <- list(Acid = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage[1:100],
                      Cis = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage[1:100],
                      CoCl2 = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage[1:100],
                      Dab = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage[1:100],
                      Tram = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage[1:100])
top100_venn <- venn(top100, intersections = T)
top100_venn_data <- attr(top100_venn,"intersections")

# Establish resistant lineages as the 200 barcodes with the highest expression
top200 <- list(Acid = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage[1:200],
                      Cis = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage[1:200],
                      CoCl2 = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage[1:200],
                      Dab = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage[1:200],
                      Tram = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage[1:200])
top200_venn <- venn(top200, intersections = T)
top200_venn_data <- attr(top200_venn,"intersections")

# Establish resistant lineages as the 500 barcodes with the highest expression
top500 <- list(Acid = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage[1:500],
                      Cis = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage[1:500],
                      CoCl2 = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage[1:500],
                      Dab = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage[1:500],
                      Tram = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage[1:500])
top500_venn <- venn(top500, intersections = T)
top500_venn_data <- attr(top500_venn,"intersections")
```

# Plot some pairwise gDNA expression between samples
```{r}

plot_df_cis_acid <- merge(gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt,gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt, by = 'Lineage', all = T)
ggplot(plot_df_cis_acid, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Cis vs Acid - Spearman:'), cor(plot_df_cis_acid$Reads.x,plot_df_cis_acid$Reads.y, method = 'spearman'))  

plot_df_cocl2_acid <- merge(gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt,gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt, by = 'Lineage', all = T)
ggplot(plot_df_cocl2_acid, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('CoCl2 vs Acid - Spearman:'), cor(plot_df_cocl2_acid$Reads.x,plot_df_cocl2_acid$Reads.y, method = 'spearman'))

plot_df_dab_acid <- merge(gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt,gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt, by = 'Lineage', all = T)
ggplot(plot_df_dab_acid, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Dab vs Acid - Spearman:'), cor(plot_df_dab_acid$Reads.x,plot_df_dab_acid$Reads.y, method = 'spearman'))

plot_df_tram_acid <- merge(gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt,gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt, by = 'Lineage', all = T)
ggplot(plot_df_tram_acid, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Tram vs Acid - Spearman:'), cor(plot_df_tram_acid$Reads.x,plot_df_tram_acid$Reads.y, method = 'spearman'))

plot_df_tram_dab <- merge(gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt, gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt, by = 'Lineage', all = T)
ggplot(plot_df_tram_dab, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Tram vs Dab - Spearman:'), cor(plot_df_tram_acid$Reads.x,plot_df_dab_acid$Reads.y, method = 'spearman'))

plot_df_cis_dab <- merge(gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt, gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt, by = 'Lineage', all = T)
ggplot(plot_df_cis_dab, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Cis vs Dab - Spearman:'), cor(plot_df_cis_dab$Reads.x,plot_df_cis_dab$Reads.y, method = 'spearman'))

plot_df_cocl2_dab <- merge(gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt, gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt, by = 'Lineage', all = T)
ggplot(plot_df_cocl2_dab, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('CoCl2 vs Dab - Spearman:'), cor(plot_df_cocl2_dab$Reads.x,plot_df_cocl2_dab$Reads.y, method = 'spearman'))

plot_df_cocl2_tram <- merge(gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt, gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt, by = 'Lineage', all = T)
ggplot(plot_df_cocl2_tram, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('CoCl2 vs Tram - Spearman:'), cor(plot_df_cocl2_tram$Reads.x,plot_df_cocl2_tram$Reads.y, method = 'spearman'))

plot_df_cis_tram <- merge(gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt, gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt, by = 'Lineage', all = T)
ggplot(plot_df_cis_tram, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('Cis vs Tram - Spearman:'), cor(plot_df_cis_tram$Reads.x,plot_df_cis_tram$Reads.y, method = 'spearman'))

plot_df_cocl2_cis <- merge(gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt, gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt, by = 'Lineage', all = T)
ggplot(plot_df_cocl2_cis, mapping = aes(x = log2(Reads.x), y = log2(Reads.y))) + geom_point() + ggtitle(paste0('CoCl2 vs Cis - Spearman:'), cor(plot_df_cocl2_cis$Reads.x,plot_df_cocl2_cis$Reads.y, method = 'spearman'))

```

# Start loading in the scRNA-seq data
```{r include=FALSE}
cocl2.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_CoCl2_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
acid.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Acid_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
dab.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Dab_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
tram.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Tram_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
cis.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Cis_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
naive1.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Naive1_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
naive2.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Naive2_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
naive3.data <- Read10X(data.dir = '/Volumes/Super_cells/2021_10_01_United_gDNA_10xcDNA_BCs_RunCellRanger/DLS005_Naive3_BC_count_fromUnited/outs/filtered_feature_bc_matrix')
```

# Convert gene expression into a seurat object
```{r}
rm(temp, temp2,temp3)

cocl2 <- CreateSeuratObject(counts = cocl2.data$`Gene Expression`)
acid <- CreateSeuratObject(counts = acid.data$`Gene Expression`)
dab <- CreateSeuratObject(counts = dab.data$`Gene Expression`)
tram <- CreateSeuratObject(counts = tram.data$`Gene Expression`)
cis <- CreateSeuratObject(counts = cis.data$`Gene Expression`)
naive1 <- CreateSeuratObject(counts = naive1.data$`Gene Expression`)
naive2 <- CreateSeuratObject(counts = naive2.data$`Gene Expression`)
naive3 <- CreateSeuratObject(counts = naive3.data$`Gene Expression`)
```

# Add lineage data to the seurat object
```{r}
cocl2[['lineage']] <- CreateAssayObject(counts = cocl2.data$Custom, min.cells = 1)
acid[['lineage']] <- CreateAssayObject(counts = acid.data$Custom, min.cells = 1)
dab[['lineage']] <- CreateAssayObject(counts = dab.data$Custom, min.cells = 1)
tram[['lineage']] <- CreateAssayObject(counts = tram.data$Custom, min.cells = 1)
cis[['lineage']] <- CreateAssayObject(counts = cis.data$Custom, min.cells = 1)
naive1[['lineage']] <- CreateAssayObject(counts = naive1.data$Custom, min.cells = 1)
naive2[['lineage']] <- CreateAssayObject(counts = naive2.data$Custom, min.cells = 1)
naive3[['lineage']] <- CreateAssayObject(counts = naive3.data$Custom, min.cells = 1)
```

# Determine the amount of mitochondrial and ribosomal genes in each condition --------
```{r}
cocl2[["percent.mt"]] <- PercentageFeatureSet(object = cocl2, pattern = "^MT-")
cocl2[["percent.rb"]] <- PercentageFeatureSet(object = cocl2, pattern = "^RPS")

acid[["percent.mt"]] <- PercentageFeatureSet(object = acid, pattern = "^MT-")
acid[["percent.rb"]] <- PercentageFeatureSet(object = acid, pattern = "^RPS")

dab[["percent.mt"]] <- PercentageFeatureSet(object = dab, pattern = "^MT-")
dab[["percent.rb"]] <- PercentageFeatureSet(object = dab, pattern = "^RPS")

tram[["percent.mt"]] <- PercentageFeatureSet(object = tram, pattern = "^MT-")
tram[["percent.rb"]] <- PercentageFeatureSet(object = tram, pattern = "^RPS")

cis[["percent.mt"]] <- PercentageFeatureSet(object = cis, pattern = "^MT-")
cis[["percent.rb"]] <- PercentageFeatureSet(object = cis, pattern = "^RPS")

naive1[["percent.mt"]] <- PercentageFeatureSet(object = naive1, pattern = "^MT-")
naive1[["percent.rb"]] <- PercentageFeatureSet(object = naive1, pattern = "^RPS")

naive2[["percent.mt"]] <- PercentageFeatureSet(object = naive2, pattern = "^MT-")
naive2[["percent.rb"]] <- PercentageFeatureSet(object = naive2, pattern = "^RPS")

naive3[["percent.mt"]] <- PercentageFeatureSet(object = naive3, pattern = "^MT-")
naive3[["percent.rb"]] <- PercentageFeatureSet(object = naive3, pattern = "^RPS")
```

#Visualize QC metrics as a violin plot
```{r}
VlnPlot(object = cocl2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = cocl2, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = cocl2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = acid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = acid, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = acid, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = dab, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = dab, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = dab, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = tram, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = tram, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = tram, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = cis, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = cis, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = cis, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = naive1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = naive1, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = naive1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = naive2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = naive2, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = naive2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(object = naive3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
FeatureScatter(object = naive3, feature1 = "nCount_RNA", feature2 = "percent.mt") 
FeatureScatter(object = naive3, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

# From QC plots above set filter parameters, and replot after filtration
```{r}
#subset based on filter parameters
cocl2_onc <- cocl2@assays$lineage@counts@Dim[2]
cocl2 <- subset(x = cocl2, subset = nFeature_RNA > 2500 & nCount_RNA < 40000 & percent.mt < 15) # percent.mt < 25
VlnPlot(object = cocl2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
cocl2_f1 <- cocl2@assays$lineage@counts@Dim[2]/cocl2_onc
print(cocl2_f1*100)

#subset based on filter parameters
acid_onc <- acid@assays$lineage@counts@Dim[2]
acid <- subset(x = acid, subset = nFeature_RNA > 1500 & nCount_RNA < 25000 & percent.mt < 15) #  
VlnPlot(object = acid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
acid_f1 <- acid@assays$lineage@counts@Dim[2]/acid_onc
print(acid_f1*100)

#subset based on filter parameters
dab_onc <- dab@assays$lineage@counts@Dim[2]
dab <- subset(x = dab, subset = nFeature_RNA > 2000 & nCount_RNA < 25000 & percent.mt < 15) #  
VlnPlot(object = dab, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
dab_f1 <- dab@assays$lineage@counts@Dim[2]/dab_onc
print(dab_f1*100)

#subset based on filter parameters
tram_onc <- tram@assays$lineage@counts@Dim[2]
tram <- subset(x = tram, subset = nFeature_RNA > 2000 & nCount_RNA < 20000 & percent.mt < 15) #  
VlnPlot(object = tram, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
tram_f1 <- tram@assays$lineage@counts@Dim[2]/tram_onc
print(tram_f1*100)

#subset based on filter parameters
cis_onc <- cis@assays$lineage@counts@Dim[2]
cis <- subset(x = cis, subset = nFeature_RNA > 1500 & nCount_RNA < 20000 & percent.mt < 20) #
VlnPlot(object = cis, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
cis_f1 <- cis@assays$lineage@counts@Dim[2]/cis_onc
print(cis_f1*100)

#subset based on filter parameters
naive1_onc <- naive1@assays$lineage@counts@Dim[2]
naive1 <- subset(x = naive1, subset = nFeature_RNA > 3000 & nCount_RNA < 55000 & percent.mt < 20) # 
VlnPlot(object = naive1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
naive1_f1 <- naive1@assays$lineage@counts@Dim[2]/naive1_onc
print(naive1_f1*100)

#subset based on filter parameters
naive2_onc <- naive2@assays$lineage@counts@Dim[2]
naive2 <- subset(x = naive2, subset = nFeature_RNA > 3000 & nCount_RNA < 50000 & percent.mt < 20) # 
VlnPlot(object = naive2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
naive2_f1 <- naive2@assays$lineage@counts@Dim[2]/naive2_onc
print(naive2_f1*100)

#subset based on filter parameters
naive3_onc <- naive3@assays$lineage@counts@Dim[2]
naive3 <- subset(x = naive3, subset = nFeature_RNA > 2500 & nCount_RNA < 50000 & percent.mt < 15) # 
VlnPlot(object = naive3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), ncol = 4)
#percent of cells kept after this filter
naive3_f1 <- naive3@assays$lineage@counts@Dim[2]/naive3_onc
print(naive3_f1*100)
```

# Normalize and identify variable features
```{r include=FALSE}
# Remove unnecessary files
rm(list = c('acid.data',
            'cis.data',
            'cocl2.data',
            'dab.data',
            'tram.data',
            'naive1.data',
            'naive2.data',
            'naive3.data'))

# allow R to use more memory
options(future.globals.maxSize= 35 * 1024^2)

cocl2 <- SCTransform(cocl2, verbose = FALSE)
acid <- SCTransform(acid, verbose = FALSE)
dab <- SCTransform(dab, verbose = FALSE)
tram <- SCTransform(tram, verbose = FALSE)
cis <- SCTransform(cis, verbose = FALSE)
naive1 <- SCTransform(naive1, verbose = FALSE)
naive2 <- SCTransform(naive2, verbose = FALSE)
naive3 <- SCTransform(naive3, verbose = FALSE)
```

# Add cell cycle scores to cells 
```{r include=FALSE}
cocl2 <- CellCycleScoring(object = cocl2, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
acid <- CellCycleScoring(object = acid, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
dab <- CellCycleScoring(object = dab, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
tram <- CellCycleScoring(object = tram, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
cis <- CellCycleScoring(object = cis, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
naive1 <- CellCycleScoring(object = naive1, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
naive2 <- CellCycleScoring(object = naive2, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
naive3 <- CellCycleScoring(object = naive3, g2m.features = cc.genes$g2m.genes, s.features = cc.genes$s.genes)
```

# Run PCA and visulize results
```{r include=FALSE}
cocl2 <- RunPCA(object = cocl2)
acid <- RunPCA(object = acid)
dab <- RunPCA(object = dab)
tram <- RunPCA(object = tram)
cis <- RunPCA(object = cis)
naive1 <- RunPCA(object = naive1)
naive2 <- RunPCA(object = naive2)
naive3 <- RunPCA(object = naive3)
```

# Plot heatmap of first PC
```{r}
DimHeatmap(object = cocl2, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = acid, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = dab, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = tram, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = cis, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = naive1, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = naive2, dims = 1, cells = 10000, balanced = TRUE)
DimHeatmap(object = naive3, dims = 1, cells = 10000, balanced = TRUE)
```

#ploting known markers of primed and non-resistant state
```{r}
FeaturePlot(object = cocl2, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = acid, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = dab, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = tram, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = cis, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = naive1, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = naive2, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
FeaturePlot(object = naive3, features = c("NGFR","EGFR","AXL","FN1","MITF"), reduction = "pca" ,pt.size = 1, combine = FALSE, order = TRUE)
```

# Run UMAP
```{r include=FALSE}
cocl2 <- FindNeighbors(object = cocl2, dims = c(1:20))
cocl2 <- FindClusters(object = cocl2, resolution = .5)
cocl2 <- RunUMAP(object = cocl2, dims = c(1:20))

acid <- FindNeighbors(object = acid, dims = c(1:20))
acid <- FindClusters(object = acid, resolution = .5)
acid <- RunUMAP(object = acid, dims = c(1:20))

dab <- FindNeighbors(object = dab, dims = c(1:20))
dab <- FindClusters(object = dab, resolution = .5)
dab <- RunUMAP(object = dab, dims = c(1:20))

tram <- FindNeighbors(object = tram, dims = c(1:20))
tram <- FindClusters(object = tram, resolution = .5)
tram <- RunUMAP(object = tram, dims = c(1:20))

cis <- FindNeighbors(object = cis, dims = c(1:20))
cis <- FindClusters(object = cis, resolution = .5)
cis <- RunUMAP(object = cis, dims = c(1:20))

naive1 <- FindNeighbors(object = naive1, dims = c(1:20))
naive1 <- FindClusters(object = naive1, resolution = .5)
naive1 <- RunUMAP(object = naive1, dims = c(1:20))

naive2 <- FindNeighbors(object = naive2, dims = c(1:20))
naive2 <- FindClusters(object = naive2, resolution = .5)
naive2 <- RunUMAP(object = naive2, dims = c(1:20))

naive3 <- FindNeighbors(object = naive3, dims = c(1:20))
naive3 <- FindClusters(object = naive3, resolution = .5)
naive3 <- RunUMAP(object = naive3, dims = c(1:20))
```

# Plot UMAP
```{r}
#show clusters in UMAP
DimPlot(cocl2, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters CoCl2')
DimPlot(acid, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Acid')
DimPlot(dab, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Dabrafenib')
DimPlot(tram, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Trametinib')
DimPlot(cis, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Cisplatin')
DimPlot(naive1, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Naive1')
DimPlot(naive2, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Naive2')
DimPlot(naive3, reduction = "umap",dims = c(1,2), group.by = 'seurat_clusters', pt.size = 1) + 
  DarkTheme() + ggtitle('Seurat Clusters Naive3')
```

# Make sure that the same 10x barcodes arent found across multiple samples (a small level of overlap is okay, the same barcodes are in each tube. Just don't want complete overlap)
```{r}
naive1_names <- colnames(naive1)
naive2_names <- colnames(naive2)
naive3_names <- colnames(naive3)

naive1_2_int <- length(intersect(naive1_names, naive2_names))
naive1_3_int <- length(intersect(naive1_names, naive3_names))
naive2_3_int <- length(intersect(naive2_names, naive3_names))

acid_names <- colnames(acid)
cocl2_names <- colnames(cocl2)
dab_names <- colnames(dab)
tram_names <- colnames(tram)
cis_names <- colnames(cis)

acid_cocl2_int <- length(intersect(acid_names, cocl2_names))
acid_dab_int <- length(intersect(acid_names, dab_names))
acid_tram_int <- length(intersect(acid_names, tram_names))
acid_cis_int <- length(intersect(acid_names,cis_names))

cocl2_dab_int <- length(intersect(cocl2_names, dab_names))
cocl2_tram_int <- length(intersect(cocl2_names, tram_names))
cocl2_cis_int <- length(intersect(cocl2_names,cis_names))

dab_tram_int <- length(intersect(dab_names, tram_names))
dab_cis_int <- length(intersect(dab_names,cis_names))

tram_cis_int <- length(intersect(tram_names,cis_names))

```

# Need to switch back to RNA assay and perform Normalization and scaling for DE or plotting genes as suggested here
```{r include=FALSE}
DefaultAssay(object = cocl2) <- "RNA"
cocl2 <- NormalizeData(cocl2)
cocl2 <- ScaleData(cocl2)

DefaultAssay(object = acid) <- "RNA"
acid <- NormalizeData(acid)
acid <- ScaleData(acid)

DefaultAssay(object = dab) <- "RNA"
dab <- NormalizeData(dab)
dab <- ScaleData(dab)

DefaultAssay(object = tram) <- "RNA"
tram <- NormalizeData(tram)
tram <- ScaleData(tram)

DefaultAssay(object = cis) <- "RNA"
cis <- NormalizeData(cis)
cis <- ScaleData(cis)

DefaultAssay(object = naive1) <- "RNA"
naive1 <- NormalizeData(naive1)
naive1 <- ScaleData(naive1)

DefaultAssay(object = naive2) <- "RNA"
naive2 <- NormalizeData(naive2)
naive2 <- ScaleData(naive2)

DefaultAssay(object = naive3) <- "RNA"
naive3 <- NormalizeData(naive3)
naive3 <- ScaleData(naive3)
```

# Plot marker genes on UMAP
```{r}
FeaturePlot(object = cocl2, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = acid, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = dab, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = tram, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = cis, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = naive1, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = naive2, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
FeaturePlot(object = naive3, reduction = 'umap',features = c("NGFR","EGFR","AXL","NT5E","FN1","SERPINE2","A2M","MITF","SOX10"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
```

# Save and load the the workspace - premerging
```{r}
save.image(file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/workspace_premerge.RData')
#load('/Volumes/Super_cells/2021_10_28_Cleaned_Data/workspace_premerge.RData')
```

# Merge the datasets
```{r}
all_naive <- merge(naive1, y = c(naive2, naive3), add.cell.ids = c("naive1", "naive2", "naive3"), project = "Naive")
all_resistant <- merge(cocl2, y = c(acid, dab, tram, cis), add.cell.ids = c("cocl2", "acid", "dab", "tram", "cis"), project = "Resistant")
all_data <- merge(all_naive, y = all_resistant, add.cell.ids = c("all_naive", "all_resistant"), project = "All Data")
```

# Remove the excess seurat objects and save a "post-merge"
```{r}
rm(acid,cis,cocl2,dab,tram,naive1,naive2,naive3,plot_list,plot_list2)   
save(all_naive, all_resistant, all_data, file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/objects_postmerge.RData')
#load('/Volumes/Super_cells/2021_10_28_Cleaned_Data/objects_postmerge.RData')
```

# SCTransform the different Seurat objects
```{r include=FALSE}
# allow R to use more memory
options(future.globals.maxSize= 35 * 1024^2)

all_naive <- SCTransform(all_naive, verbose = F)
save(all_naive, file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/all_naive_SCT.RData')
rm('all_naive')

all_resistant <- SCTransform(all_resistant, verbose = F)
save(all_resistant, file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/all_resistant_SCT.RData')
rm('all_resistant')

all_data <- SCTransform(all_data, verbose = F)
save(all_data, file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/all_data_SCT.RData')
```

# Plot UMAPs of all data
```{r}
condition <- c()
for (i in 1:length(all_data@assays$RNA@counts@Dimnames[2][[1]])){
  condition <- rbind(condition, strsplit(all_data@assays$RNA@counts@Dimnames[2][[1]][[i]], '_')[[1]][3])
}

all_data$OG_condition <- condition
all_data$OG_condition[all_data$OG_condition %in% c('naive1', 'naive2', 'naive3')] <- 'naive'
all_data <- RunPCA(all_data, verbose = F)
all_data <- RunUMAP(all_data, dims = 1:30, verbose = F)
all_data <- FindNeighbors(all_data, dims = 1:30, verbose = F)
all_data <- FindClusters(all_data, verbose = F)
DimPlot(all_data, label = T) + DarkTheme()
FeaturePlot(object = all_data, reduction = 'umap',features = c("AXL","EGFR","NT5E","MITF"), pt.size = 1, combine = T, order = TRUE) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
DimPlot(all_data, reduction = 'umap', group.by = 'OG_condition') 



save(all_data, file = '/Volumes/Super_cells/2021_10_28_Cleaned_Data/all_data_SCT.RData')
```

# Plot the all_data umap with each condition in its color separately
```{r}
# Plotting overlay
cocl2_res <- names(all_data$orig.ident[all_data$OG_condition == 'cocl2']) # Gets cocl2 res cells
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(cocl2_res),
        cols.highlight = 'forestgreen') + DarkTheme() + ggtitle('CoCl2')

acid_res <- names(all_data$orig.ident[all_data$OG_condition == 'acid']) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(acid_res),
        cols.highlight = 'darksalmon') + DarkTheme() + ggtitle('Acid')

dab_res <- names(all_data$orig.ident[all_data$OG_condition == 'dab']) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(dab_res),
        cols.highlight = 'cyan') + DarkTheme() + ggtitle('Dabrafenib')

tram_res <- names(all_data$orig.ident[all_data$OG_condition == 'tram']) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(tram_res),
        cols.highlight = 'hotpink') + DarkTheme() + ggtitle('Trametinib')

cis_res <- names(all_data$orig.ident[all_data$OG_condition == 'cis']) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(cis_res),
        cols.highlight = 'darkgoldenrod') + DarkTheme() + ggtitle('Cisplatin')

naive_res <- names(all_data$orig.ident[all_data$OG_condition == 'naive']) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'OG_condition', pt.size = .1, cells.highlight = list(naive_res),
        cols.highlight = '#00A9FF') + DarkTheme() + ggtitle('Naive')
```

# Remove lineages that only have a single read across all cells and that are not in the updated reference file
```{r}
# Remove lineages that only have <= 1 count across all of the cells
counts <- GetAssayData(all_data, assay = 'lineage')
num_lin_orig <- nrow(counts)
counts_filt <- counts[rownames(counts) %in% reference$name,]

counts_filt <- counts_filt[which(rowSums(counts_filt) > 1),] # Only care about lineages that have greater than one total count
num_lin_filt1 <- nrow(counts_filt)
percent_lin_filt1 <- num_lin_filt1/num_lin_orig*100

df_counts <- as.data.frame(counts_filt)

# Find how many barcodes each cell has that has greater than 0 counts
barcodes_per_cell <- apply(counts_filt,2, function(x) sum(x>0))
table2 <- table(barcodes_per_cell)
```

# Try doing lineage count filtering on a lane by lane basis (would want to split the naive into 3 here)
```{r include=FALSE}
condition <- c()
for (i in 1:length(all_data@assays$RNA@counts@Dimnames[2][[1]])){
  condition <- rbind(condition, strsplit(all_data@assays$RNA@counts@Dimnames[2][[1]][[i]], '_')[[1]][3])
}
all_data$OG_condition_naivesplit <- condition


# Pipe this to run as a for loop through each of the different conditions
plot_df_list <- list() 

for (k in unique(all_data$OG_condition_naivesplit)){
  print(k)
plot_df <- data.frame()
for (i in 1:50){ # 1:20
  temp_fam_list <- list()
  print(i)
  count <- 1
  for (j in grep(k,colnames(counts_filt))){
    temp_fam_list[[count]] <- which(counts_filt[,j] > i)
    count <- count+1
  }
  table <- table(lengths(temp_fam_list))
  temp <- data.frame(zero = as.numeric(table[1]), one = as.numeric(table[2]), two_or_more = length(temp_fam_list) - as.numeric(table[1]) - as.numeric(table[2]))
  plot_df <- rbind(plot_df,temp)
}
plot_df_list[[k]] <- plot_df
}

plot_df_list2 <- list()
for (i in names(plot_df_list)){
  plot_df_list2[[i]] <- data.frame(
  count_filt = rep(1:50,3),
  num_cells = unlist(plot_df_list[[i]]),
  barcode_num = c(rep('zero',50),rep('one',50), rep('multiple',50))
)
}

```

# Show the plots
```{r}

naive1_BC_filt <- ggplot(data = plot_df_list2[[1]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[1])
naive1_BC_filt

naive2_BC_filt <- ggplot(data = plot_df_list2[[2]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[2])
naive2_BC_filt

naive3_BC_filt <- ggplot(data = plot_df_list2[[3]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[3])
naive3_BC_filt

cocl2_BC_filt <- ggplot(data = plot_df_list2[[4]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[4])
cocl2_BC_filt

acid_BC_filt <- ggplot(data = plot_df_list2[[5]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[5])
acid_BC_filt

dab_BC_filt <- ggplot(data = plot_df_list2[[6]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[6])
dab_BC_filt

tram_BC_filt <- ggplot(data = plot_df_list2[[7]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[7])
tram_BC_filt

cis_BC_filt <- ggplot(data = plot_df_list2[[8]], aes(x = count_filt, y = num_cells, group = barcode_num, color = barcode_num)) + geom_point()+geom_line() + labs(title = names(plot_df_list2)[8])
cis_BC_filt
```

# Filter based on condition-by-condition thresholds
```{r}
# Find best way to set the threshold by trying a couple of different metrics
cutoffs <-data.frame()
for(i in names(plot_df_list2)){
  zero <- plot_df_list2[[i]][plot_df_list2[[i]]$barcode_num == 'zero',]
  one <- plot_df_list2[[i]][plot_df_list2[[i]]$barcode_num == 'one',]
  multiple <- plot_df_list2[[i]][plot_df_list2[[i]]$barcode_num == 'multiple',]
  cutoffs <- rbind(cutoffs, data.frame(Max_single = which.max(one$num_cells), Max_ratio = which.max(one$num_cells/multiple$num_cells), Max_difference = which.max(one$num_cells - multiple$num_cells)))
}
rownames(cutoffs) <- names(plot_df_list2)

# Set the filter thresholds as the count that maximizes the number of cells with one barcode - cells with multiple barcodes
filter_thresholds <- cutoffs$Max_difference
names(filter_thresholds) <- names(plot_df_list)

# filter based on decided threshold
fam <- list()
for (k in names(filter_thresholds)){
  for (i in grep(k,colnames(df_counts))){
    fam[[i]] <- rownames(df_counts)[which(df_counts[,i] > filter_thresholds[k])]
    }
}
# See how many cells still have reads from how many barcodes
table_fam <- table(lengths(fam))

# Build a boolean of the cells that still have multiple barcodes
boolean_multi_lin <- lengths(fam)>1

# Assign dominant barcode if the number of counts in the highest expressed barcodes is >triple that of second highest barcode
fixed_multi <- list()
for (i in (1:ncol(df_counts))){ 
  if ( boolean_multi_lin[i] == T) {
      if(df_counts[topn(as.numeric(df_counts[,i]),2),i][1] >= 3*df_counts[topn(as.numeric(df_counts[,i]),2),i][2]){
        fixed_multi[[i]] <- rownames(df_counts)[topn(as.numeric(df_counts[,i]),2)][1]
      }else{
        fixed_multi[[i]] <- "Still multiple"
      }
  } else{ fixed_multi[[i]] <- "Single"
}
}

# Get just the naive cells
Idents(all_data) <- all_data$OG_condition
only_naive <- WhichCells(all_data, idents = all_data$OG_condition[all_data$OG_condition %in% 'naive'])
`%nin%` = Negate(`%in%`)

# Replace the values in fam that had 2+ barcodes with with either the dominant barcode if found or say that its still multiple
fam_multi_collapse <- fam
fam_multi_collapse[boolean_multi_lin] <- fixed_multi[boolean_multi_lin]

# Calulcalate how many cells still have multuple barcodes and how many have just one
number_fam_mult <- length(fam_multi_collapse[fam_multi_collapse == 'Still multiple'])
number_fam_single <- ncol(df_counts) - number_fam_mult - table_fam[1]

# Give cells without a barcode the label 'No barcode'
fam_multi_collapse[lengths(fam_multi_collapse) == 0] <- 'No Barcode'

# Format into metadata so that final lineage determinations can be added to the seurat object
lin_meta <- t(as.data.frame(fam_multi_collapse))
colnames(lin_meta) <- 'final_lineage'
rownames(lin_meta) <- colnames(df_counts)

all_data$final_lineage <- lin_meta

# Get how many cells per lineage with this method in the naive cells 
naive_cells_per_lineage_max_counts  <- as.data.frame(table(all_data$final_lineage[names(all_data$final_lineage) %in% only_naive & all_data$final_lineage %nin% c('No Barcode','Still multiple')]))

ggplot(data = naive_cells_per_lineage_max_counts, aes(reorder(Var1,-Freq),Freq))+geom_bar(stat='identity') +  
  theme(axis.text.x=element_blank()) + labs(x = 'Rank ordered lineages', y = 'Cells per lineage') 

# Make list of lineages <= threshold cells
lin_size_thresh <-10

naive_greater <- as.character(naive_cells_per_lineage_max_counts$Var1[naive_cells_per_lineage_max_counts$Freq > lin_size_thresh])
naive_lessEqual <- as.character(naive_cells_per_lineage_max_counts$Var1[naive_cells_per_lineage_max_counts$Freq <= lin_size_thresh])

# Lets see how many naive cells were in these large lineages
only_naive_largeLin <- names(all_data$final_lineage[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% naive_greater])
only_naive_smallLin <- names(all_data$final_lineage[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% naive_lessEqual])
# The few large lineages make up more cells than the many small lineages

# Make cells in the large lineage into a "too large Naive" 
all_data$final_lineage[all_data$final_lineage %in% naive_greater] <- "Too large Naive"
```

# Look at size of lineages after filtering the crazy large naive lineages
```{r}
# Bring the SCTransformed naive data back in
load('/Volumes/Super_cells/2021_10_28_Cleaned_Data/all_naive_SCT.RData')

# Need to add the lineage metadata to the naive seurat object
all_naive$final_lineage <- all_data$final_lineage[grep('naive',names(all_data$final_lineage))]
all_naive <- RunPCA(all_naive, verbose = F)
all_naive <- RunUMAP(all_naive, dims = 1:30, verbose = F)
all_naive <- FindNeighbors(all_naive, dims = 1:30, verbose = F)
all_naive <- FindClusters(all_naive, verbose = F)

# Do naive first
naive_cells_per_lin_for_plot_df <- data.frame(Lineage = all_naive$final_lineage)

temp_df <- data.frame()
for (i in names(all_naive$final_lineage)){
  if(is_empty(naive_cells_per_lineage_max_counts$Freq[naive_cells_per_lineage_max_counts$Var1 == all_naive$final_lineage[i]])){
    temp_df <- rbind(temp_df, 0)
  }else{
  temp_df <- rbind(temp_df, naive_cells_per_lineage_max_counts$Freq[naive_cells_per_lineage_max_counts$Var1 == all_naive$final_lineage[i]])
  }
}
naive_cells_per_lin_for_plot_df$Freq <- temp_df$X2L

all_naive$Size_Lin <- naive_cells_per_lin_for_plot_df$Freq
FeaturePlot(all_naive, features = 'Size_Lin') & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))

# Now do all of the data
# Get how many cells per lineage with this method
all_cells_per_lineage_max_counts  <- as.data.frame(table(all_data$final_lineage[all_data$final_lineage %nin% c('No Barcode','Still multiple', 'Too large Naive')]))

all_cells_per_lin_for_plot_df <- data.frame(Lineage = all_data$final_lineage)

temp_df2 <- data.frame()
for (i in names(all_data$final_lineage)){
  if(is_empty(all_cells_per_lineage_max_counts$Freq[all_cells_per_lineage_max_counts$Var1 == all_data$final_lineage[i]])){
    temp_df2 <- rbind(temp_df2, 0)
  }else{
  temp_df2 <- rbind(temp_df2, all_cells_per_lineage_max_counts$Freq[all_cells_per_lineage_max_counts$Var1 == all_data$final_lineage[i]])
  }
}
all_cells_per_lin_for_plot_df$Freq <- temp_df2$X4L

all_data$Size_Lin <- all_cells_per_lin_for_plot_df$Freq
FeaturePlot(all_data, features = 'Size_Lin') & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))
```

# Lets try to look at barcodes in the naive cells with new filtering
```{r}
FeaturePlot(all_naive, features = c('SOX10','MITF','FN1','AXL','EGFR','NT5E')) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))

# Plotting overlay top 200
cocl2_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$CoCl2]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(cocl2_res_cells_naive_only),
        cols.highlight = 'forestgreen') + DarkTheme() + ggtitle('CoCl2')

acid_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$Acid]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(acid_res_cells_naive_only),
        cols.highlight = 'darksalmon') + DarkTheme() + ggtitle('Acid')

dab_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$Dab]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(dab_res_cells_naive_only),
        cols.highlight = 'cyan') + DarkTheme() + ggtitle('Dabrafenib')

tram_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$Tram]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(tram_res_cells_naive_only),
        cols.highlight = 'hotpink') + DarkTheme() + ggtitle('Trametinib')

cis_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$Cis]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(cis_res_cells_naive_only),
        cols.highlight = 'darkgoldenrod') + DarkTheme() + ggtitle('Cisplatin')

super_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$`Acid:Cis:CoCl2:Dab:Tram`]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(super_res_cells_naive_only),
        cols.highlight = 'red') + DarkTheme() + ggtitle('Super Cells')

dabtram_res_cells_naive_only <- names(all_naive$orig.ident[all_naive$final_lineage %in% top200_venn_data$`Dab:Tram`]) # Gets resistant cell identifiers
DimPlot(all_naive, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(dabtram_res_cells_naive_only),
        cols.highlight = 'orange') + DarkTheme() + ggtitle('Dab/Tram Resistant Cells')
```

# Lets try to look at barcodes in all of the cells with new filtering
```{r}
FeaturePlot(all_data, features = c('SOX10','MITF','FN1','AXL','EGFR','NT5E')) & 
  DarkTheme() & scale_colour_gradientn(colors = rev(brewer.pal(n = 11, name = 'RdBu')))

# CoCl2
cocl2_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$CoCl2]) # Gets resistant cell identifiers
cocl2_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$CoCl2]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(cocl2_res_cells),
        cols.highlight = 'forestgreen') + DarkTheme() + ggtitle('CoCl2 top 200')


#Acid
acid_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$Acid]) # Gets resistant cell identifiers
acid_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$Acid]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(acid_res_cells),
        cols.highlight = 'darksalmon') + DarkTheme() + ggtitle('Acid top 200')


#Dabrafenib
dab_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$Dab]) # Gets resistant cell identifiers
dab_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$Dab]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(dab_res_cells),
        cols.highlight = 'cyan') + DarkTheme() + ggtitle('Dabrafenib top 200')

#Trametinib
tram_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$Tram]) # Gets resistant cell identifiers
tram_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$Tram]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(tram_res_cells),
        cols.highlight = 'hotpink') + DarkTheme() + ggtitle('Trametinib top 200')

#Cisplatin
cis_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$Cis]) # Gets resistant cell identifiers
cis_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$Cis]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(cis_res_cells),
        cols.highlight = 'darkgoldenrod') + DarkTheme() + ggtitle('Cisplatin top 200')

#Supercells
super_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$`Acid:Cis:CoCl2:Dab:Tram`]) # Gets resistant cell identifiers
super_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$`Acid:Cis:CoCl2:Dab:Tram`]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(super_res_cells),
        cols.highlight = 'red') + DarkTheme() + ggtitle('Super Cells top 200')

#Supercells
dabtram_res_cells_naive <- names(all_data$OG_condition[all_data$OG_condition %in% 'naive' & all_data$final_lineage %in% top200_venn_data$`Dab:Tram`]) # Gets resistant cell identifiers
dabtram_res_cells <- names(all_data$OG_condition[all_data$final_lineage %in% top200_venn_data$`Dab:Tram`]) # Gets resistant cell identifiers
DimPlot(all_data, reduction = "umap",dims = c(1,2),
        group.by = 'final_lineage', pt.size = .1, cells.highlight = list(dabtram_res_cells),
        cols.highlight = 'red') + DarkTheme() + ggtitle('Dab/Tram Cells top 200')

```

# Scatter plot the gDNA reads vs cells with barcode (cells with barcode from the starcode output, so not filtered)
```{r}
plot_df_cocl2 <- data.frame(Lineage = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Lineage,
                            Reads = gDNA_anno$sc_output_counts_DLS005_CoCl2_gDNA_S35.txt$Reads)
plot_df_cocl2 <-plot_df_cocl2[plot_df_cocl2$Reads > 2,]

temp_df <- data.frame()
for (i in plot_df_cocl2$Lineage){
  temp_df <- rbind(temp_df, length(all_data$OG_condition[all_data$OG_condition %in% 'cocl2' & all_data$final_lineage %in% i]))
}
plot_df_cocl2$Single_cells <- unlist(temp_df)
names(plot_df_cocl2) <-  c('Lineage','Reads','Num_cells')
ggplot(plot_df_cocl2, mapping = aes(x = log2(Reads), y = log2(Num_cells+1))) + geom_point() + ggtitle('CoCl2')

plot_df_acid <- data.frame(Lineage = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Lineage,
                            Reads = gDNA_anno$sc_output_counts_DLS005_Acid_gDNA_S36.txt$Reads)
plot_df_acid <-plot_df_acid[plot_df_acid$Reads > 2,]
temp_df <- data.frame()
for (i in plot_df_acid$Lineage){
  temp_df <- rbind(temp_df, length(all_data$OG_condition[all_data$OG_condition %in% 'acid' & all_data$final_lineage %in% i]))
}
plot_df_acid$Single_cells <- unlist(temp_df)
names(plot_df_acid) <-  c('Lineage','Reads','Num_cells')
ggplot(plot_df_acid, mapping = aes(x = log2(Reads), y = log2(Num_cells+1))) + geom_point() + ggtitle('Acid')

plot_df_dab <- data.frame(Lineage = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Lineage,
                            Reads = gDNA_anno$sc_output_counts_DLS005_Dab_gDNA_S37.txt$Reads)
plot_df_dab <-plot_df_dab[plot_df_dab$Reads > 2,]
temp_df <- data.frame()
for (i in plot_df_dab$Lineage){
  temp_df <- rbind(temp_df, length(all_data$OG_condition[all_data$OG_condition %in% 'dab' & all_data$final_lineage %in% i]))
}
plot_df_dab$Single_cells <- unlist(temp_df)
names(plot_df_dab) <-  c('Lineage','Reads','Num_cells')
ggplot(plot_df_dab, mapping = aes(x = log2(Reads), y = log2(Num_cells+1))) + geom_point() + ggtitle('Dabrafenib')

plot_df_tram <- data.frame(Lineage = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Lineage,
                            Reads = gDNA_anno$sc_output_counts_DLS005_Tram_gDNA_S38.txt$Reads)
plot_df_tram <-plot_df_tram[plot_df_tram$Reads > 2,]
temp_df <- data.frame()
for (i in plot_df_tram$Lineage){
  temp_df <- rbind(temp_df, length(all_data$OG_condition[all_data$OG_condition %in% 'tram' & all_data$final_lineage %in% i]))
}
plot_df_tram$Single_cells <- unlist(temp_df)
names(plot_df_tram) <-  c('Lineage','Reads','Num_cells')
ggplot(plot_df_tram, mapping = aes(x = log2(Reads), y = log2(Num_cells+1))) + geom_point() + ggtitle('Trametinib')

plot_df_cis <- data.frame(Lineage = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Lineage,
                            Reads = gDNA_anno$sc_output_counts_DLS005_Cis_gDNA_S40.txt$Reads)
plot_df_cis <-plot_df_cis[plot_df_cis$Reads > 2,]
temp_df <- data.frame()
for (i in plot_df_cis$Lineage){
  temp_df <- rbind(temp_df, length(all_data$OG_condition[all_data$OG_condition %in% 'cis' & all_data$final_lineage %in% i]))
}
plot_df_cis$Single_cells <- unlist(temp_df)
names(plot_df_cis) <-  c('Lineage','Reads','Num_cells')
ggplot(plot_df_cis, mapping = aes(x = log2(Reads), y = log2(Num_cells+1))) + geom_point() + ggtitle('Cisplatin')
```

# Save image for further processing/sending to Zhang lab
```{r}
save.image('/Volumes/Super_cells/2021_10_28_Cleaned_Data/final_image.RData')
```

